src/poprox_recommender/default.py

```diff
@@ -83,27 +83,22 @@ def personalized_pipeline(num_slots: int, algo_params: dict[str, Any] | None = N
         ranker = topk_ranker
 
     pipeline = RecommendationPipeline(name=diversify)
+
+    # Compute embeddings
     pipeline.add(article_embedder, inputs=["candidate"], output="candidate")
     pipeline.add(article_embedder, inputs=["clicked"], output="clicked")
     pipeline.add(user_embedder, inputs=["clicked", "profile"], output="profile")
+
+    # Score and rank articles with diversification/calibration reranking
     pipeline.add(article_scorer, inputs=["candidate", "profile"], output="candidate")
-    pipeline.add(topk_ranker, inputs=["candidate", "profile"], output="ranked")
     pipeline.add(ranker, inputs=["candidate", "profile"], output="reranked")
```

Comment:
What's missing here is a mechanism for making the branch of the pipeline above return an empty list of articles when there are no clicks. One way to do that:
* Make the user embedding identically zero when there are no clicks
* Then the articles should all get zero scores
* Add a filter that removes articles with zero scores from the list

Reply:
Thoughts on this @mdekstrand?

Reply:
At the risk of a few more `if` statements (that the typechecker can hopefully help us with), I think I would be more inclined to use `None`, and then scores be NA (Pandas handles this tolerably well, NumPy is ok as well if we just use NaN as a sentinel value, which is what Pandas does under the hood at least until it's fully Arrow-based). The advantages to this:

* fast-track null compute instead of computing a lot of zeros expensively
* no data clearly distinguished from other sources of 0
* avoid risks of "testing for exactly zero" (although filtering for `>= 1.0e-6` would fix this)

Biggest downside I see: more logic to test for sentinels.

The NaN/NA design is what LensKit uses — scorers are supposed to return NA for anything they can't score, and the fallback scorer just asks each scorer in turn to score evertything left NA by the previous ones. I would be even happier with this solution if Pandas had a proper NA like R instead of repurposing NaN, but it is the Pandas way (Pandas does have masked array types, but they aren't as widely used; raw NumPy does not have masked arrays; Arrow does).

Reply:
Yeah, those are good points. I'll head that direction!